skimage/future/graph/graph_merge.py

```diff
@@ -128,10 +134,16 @@ def merge_hierarchical(labels, rag, thresh, rag_copy, in_place_merge,
             merge_func(rag, src, dst)
             new_id = rag.merge_nodes(src, dst, weight_func)
             _revalidate_node_edges(rag, new_id, edge_heap)
+            if merge_trace:
+                label_map = np.arange(labels.max() + 1)
+                for ix, (n, d) in enumerate(rag.nodes_iter(data=True)):
+                    for label in d['labels']:
+                        label_map[label] = ix
+                trace.append(label_map[labels])
```

Comment:
Ok this is crazy. =) Now that I see exactly what you're trying to do, I can offer some advice.

Firstly, saving a full image for each merge is a huge amount of storage. Given how easy it is to generate a label image from a label remapping (`label_map[labels]`), you could *just* save the `label_map`s.

Secondly, only the label maps for `src` and `dst` are changing at each iteration. Indeed, only the labels in `new_id` need to be updated in the label map; they include all the labels in `src` and `dst`. Instead of making a new `label_map` at each iteration, copy the existing map and update just the labels in `new_id`.

Thirdly, more generally you are making a *merge tree*, and the tree contains all the information to make these maps. I did this in my `viridis` "library" (in quotes because it's so simple... More like a bookshelf, really. =P). https://github.com/jni/viridis/blob/master/viridis/tree.py Look at the `get_map` function. All or parts of it could be used here.